Issue #2838: Make WhitespaceAround allow double-brace initialization.

liscju opened the pull request:
This patch makes double-brace initialization allowed, alternative would be to introduce option for that - im looking forward to opinion about that.

Regarding implementation details, unfortunately i wasnt able to implement this functionality in WhitespaceAroundCheck.isNotRelevantSituation method, because in this method when we return true(is not relevant) ast is later doesnt check at all. For example:
```java
new Properties() {{setProperty("double curly braces", "are not a style error");
        }};
```

For second left bracket in first line when we would return true from isNotRelevantSituation it wouldn't later check that the next token(setProperty) is not separated from previous token.

romani commented:
@liscju , see my comment https://github.com/checkstyle/checkstyle/issues/2838#issuecomment-176206672

romani commented:
@liscju , While I am doing code-review please generate diff reports against few projects to make sure no regression is happen.

liscju commented:
@romani I generated reports for projects: guava, spring-framework, infispan, protonpack, jOOL, RxJava and it seems there are no false positives. The diff report is here: http://liscju.github.io/issue2838/diff/checkstyle_merged.html , before report: http://liscju.github.io/issue2838/before/site/checkstyle.html , and after http://liscju.github.io/issue2838/after/site/checkstyle.html .

Another thing ,which seems kind of strange to me, is how little number of usages are of double-brace initializations across such big projects.

romani commented:
> Another thing ,which seems kind of strange to me, is how little number of use have double-brace initializations across such a big projects.

it is ok, I was not aware of this java feature too, I get to know about it only recently.

1)
Please put PR description to javadoc of methods "should............" code looks strange. For now I donot know how to rename methods to make it clear without javadoc

2) Please update javadoc/xdoc and clearly state to user that we ignore double-brace initialization.

3)
>         if (shouldCheckIfSeparatedFromPreviousToken(ast)
                && !Character.isWhitespace(line.charAt(before))) {

please move `line.charAt(before)` to variable `prevChar` that will help to make code similar to "...NetToken" conditions

4)
> shouldCheckIfSeparatedFromPreviousToken

usage of "If" in names is sign of bad name. I propose `shouldCheckSeparationFromPreviousToken`
Please fix in other names too

5)
Method order is chaotic and complicates code-review. After `visitToken` please order method declaration by order then appear/called in code. One day it will be automated by Check, and such work already in progress. For now please do this manually.

6)
>         return
                !(ast.getType() == TokenTypes.LITERAL_RETURN
                && ast.getFirstChild().getType() == TokenTypes.SEMI)
                && !isAnonymousInnerClassEnd(ast.getType(), nextChar)
                && !isPartOfDoubleBraceInitializerToOmitAfterCheck(ast);

indentation ...... real reason of indentation is too understand sequence without very attentive look at curly-braces. Please review all other complicated expression for this.

should be smth like: 
```
 return !(ast.getType() == TokenTypes.LITERAL_RETURN
                   && ast.getFirstChild().getType() == TokenTypes.SEMI)
            && !isAnonymousInnerClassEnd(ast.getType(), nextChar)
            && !isPartOfDoubleBraceInitializerToOmitAfterCheck(ast);
```

7)
>     
private static boolean shouldCheckIfSeparatedFromPreviousToken(DetailAST ast) {
        final int before = ast.getColumnNo() - 1;
        return before >= 0 && !isPartOfDoubleBraceInitializerToOmitBeforeCheck(ast);
    }

`...fromPreviosToken` vs `....ToOmitBeforeCheck` looks strange. I propose to rename `...ToOmitBeforeCheck` to `...ForPreviousToken` to make names consistent and without any new terms.

8)
reports are good. As you fix cosmetic changes , you are ok to be merged.

liscju commented:
@romani
1) DONE but i have hesitation if i have done it correctly. I included this in should* methods
```java
See <a href="https://github.com/checkstyle/checkstyle/pull/2845">
     * PR#2845</a> for more information why this function was needed.
```
and there is one more thing, after doing other fixes shouldCheckSeparationFromPreviousToken method body become just one method invocation:
```java
    /**
     * Check if it should be checked if previous token is separated from current by
     * whitespace. See <a href="https://github.com/checkstyle/checkstyle/pull/2845">
     * PR#2845</a> for more information why this function was needed.
     * @param ast current AST.
     * @return true if it should be checked if previous token is separated by whitespace,
     *      false otherwise.
     */
    private static boolean shouldCheckSeparationFromPreviousToken(DetailAST ast) {
        return !isPartOfDoubleBraceInitializerForPreviousToken(ast);
    }
```

i can inline it to the variable, but i have hesitation because:
1) before method there is more place for long comment with link
2) visitToken method seems consistent because there is later invocation of shouldCheckSeparationFromNextToken

What should i do?

2) DONE, it looks like this:
```
 * <p>This check does not flag as exception double brace initialization like:</p>
 * <pre>
 *   new Properties() {{
 *     setProperty("key", "value");
 *   }};
 * </pre>
```

3) DONE, must added checking if before is more or equals to 0, or else it would throw StringIndexOutOfBoundsException

```java
if (before >= 0) {
            char prevChar = line.charAt(before);
            if (shouldCheckSeparationFromPreviousToken(ast)
```

4) DONE, renamed accordingly

5) DONE, i tried to move it to minimize distance between usage and declaration, but i have always problem with which method should go after which. For example in class we have public methods at first, later should go private methods from public methods in order of first usage in public methods, but what about private methods used only in private methods, should they go just after first usage in private method? 

6) DONE, change identation there and in few other places where this seems reasonable

7) DONE, renamed  isPartOfDoubleBraceInitializerToOmitBeforeCheck,  isPartOfDoubleBraceInitializerToOmitAfterCheck to isPartOfDoubleBraceInitializerForPreviousToken, isPartOfDoubleBraceInitializerForNextToken 
accordingly

liscju commented:
@romani circleci failed because of internal error :
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: failed to get report for org.apache.maven.plugins:maven-checkstyle-plugin: Plugin org.apache.maven.plugins:maven-checkstyle-plugin:2.17 or one of its dependencies could not be resolved: Could not find artifact com.puppycrawl.tools:checkstyle:jar:6.16-SNAPSHOT in sevntu-maven (http://sevntu-checkstyle.github.com/sevntu.checkstyle/maven2) -> [Help 1]
```

romani commented:
@liscju ,
please rebase your PR over latest source code. CircleCI still builds 6.15 version, but we are 6.16 now.

liscju commented:
@romani rebased as reuqested

romani commented:
1)
> not flag as exception

"as violation"

2) 
> `See <a href="https://github.com/checkstyle/checkstyle/pull/2845"> PR#2845</a>`

please remove this link and copy-paste text.

3)
> i can inline it to the variable, but i have hesitation because:

lets keep it as method for now.

4)
> "ASSIGN,DIV_ASSIGN,PLUS_ASSIGN,MINUS_ASSIGN,STAR_ASSIGN,MOD_ASSIGN,SR_ASSIGN,BSR_ASSIGN,SL_ASSIGN,BXOR_ASSIGN,BOR_ASSIGN,BAND_ASSIGN

please change example to make list of tokens shorter to follow limitation of 100 characters in a line.

5)
>  but what about private methods used only in private methods, should they go just after first usage in private method? 

yes. Method ordering is not exact rules , it has few exception (never split Override and overloaded methods.) . The principle I use is minimization of scrolling to-and-fro (it does not mean "no scrolling" ) during reading the code from top to bottom.

6) 
Please regenerate diff reports and confirm that they did not change.

liscju commented:
1) DONE
2) DONE , modified PR description:
```
     * This function is needed to recognise double brace initialization as valid,
     * unfortunately its not possible to implement this functionality
     * in {@link WhitespaceAroundCheck.isNotRelevantSituation method}, because in this
     * method when we return true(is not relevant) ast is later doesnt check at all. For example:
     * new Properties() {{setProperty("double curly braces", "are not a style error");
     * }};
```
4) I don't understand what should i do
6) I run into error while trying to generate report:
```
Successfull preparation stage.
java.nio.file.NoSuchFileException: /home/liscju/Checkstyle/liscju.github.io/issue2838v2/after/site/./xref/spring-framework/spring-context/src/test/java/org/springframework/context/annotation/ComponentScanAnnotationIntegrationTests.html
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:524)
	at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:253)
	at java.nio.file.Files.copy(Files.java:1227)
	at attatrol.ahsm.FilesystemUtils.copy(FilesystemUtils.java:64)
	at attatrol.ahsm.site.XrefCopier.copyXrefFiles(XrefCopier.java:40)
	at attatrol.ahsm.Main.mergingStage(Main.java:136)
	at attatrol.ahsm.Main.main(Main.java:88)
Failed to merge file properly
```

Maybe anybody had similiar issue? I ran tests on several projects(including spring-framework which leads to error) and uses mycheck.xml with WhitespaceAround check.

romani commented:
> 4) I don't understand what should i do

Just try to make such example that will take less then 100 symbols in javadoc line.

romani commented:
@liscju , if you still have problems with tool , please try https://github.com/checkstyle/contribution/tree/master/patch-diff-report-tool it is is new version of this tool.

liscju commented:
@romani 
Trying to generate report(with checkstyle-tester) give me warnings:
Generating report with checkstyle-tester give me this:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3:18.986s
[INFO] Finished at: Sun Feb 07 21:27:37 CET 2016
[INFO] Final Memory: 40M/1093M
[INFO] ------------------------------------------------------------------------
Running Checkstyle on src/main/java - finished
linking report to index.html
Removing non refernced xref files in report ...
Backuping files that are refenced in report ...
warning: <div not found.
warning: class="section"><h2>Checkstyle not found.
warning: Results<a not found.
warning: /></a></p></div><div not found.
warning: class="section"><h2>Summary<a not found.
warning: name="Summary"></a></h2><table not found.
(many namy more)
warning: whitespace.</td>./guava/guava-gwt/src-super/com/google/common/base/super/com/google/common/base/CharMatcher.html.save not found.
Done. Result report is locates at: target/site/index.html
```

Later using patch-diff-report-tool i have NPE:
```
Successfull preparation stage.
Exception in thread "main" java.lang.NullPointerException
	at java.util.TreeMap.compare(TreeMap.java:1290)
	at java.util.TreeMap.put(TreeMap.java:538)
	at com.github.checkstyle.data.ParsedContent.addRecords(ParsedContent.java:75)
	at com.github.checkstyle.parser.StaxParserProcessor.parseXmlPortion(StaxParserProcessor.java:211)
	at com.github.checkstyle.parser.StaxParserProcessor.parse(StaxParserProcessor.java:134)
	at com.github.checkstyle.Main.executeStages(Main.java:273)
	at com.github.checkstyle.Main.main(Main.java:237)
```

Problem is with instruction(ParsedContent.addRecords in ParsedContent.java:75):
```java
final List<CheckstyleRecord> popped =
            records.put(filename, newRecords);
```

and filename, newRecords are null. 

I want to know if anybody else had the same problem, before i start to dig what is wrong. Problem appeared very recently, before i only had problems when the reports were too huge, but it was plain OutOfMemoryException and nothing else.

rnveach commented:
@liscju 
> Backuping files that are refenced in report ...
> warning: <div not found.

I had this same issue. The problem was the results of site's generated html file was all on 1 line, while other users file was split across multiple lines.
I solved this problem by removing the minimization routine which is these lines: https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch.sh#L95-L141
from `echo "Removing non refernced xref files in report ..."` to `cd ../../../`.

liscju commented:
@rnveach i tried it and the problem still occurres(NPE) , it is really weird - i have used old version of tool in the past and had no problems(beside famous OutOfMemoryException for huge report :P)

rnveach commented:
I have never used patch-diff-report-tool so I don't know what it's problem is. I just use ahsm to generate differences.

romani commented:
> Later using patch-diff-report-tool i have NPE:

please open issue on this tool https://github.com/checkstyle/contribution/issues
@attatrol , please take a look .

> Trying to generate report(with checkstyle-tester) give me warnings:

It is really strange when maven plugin decide to generate report as single line. @liscju , can you share with me some result of "mvn --version" output ?

> 4) I don't understand what should i do

@liscju , Please do this update. That will let me be ready to merge you even we have problems with reports. 
Please comment out all line in launch.sh from https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch.sh#L95 till the end.  Please share reports, do diff by https://www.diffchecker.com/ or similar and share link to diff too. While there is problem in tools we need to do this manually to let this PR be merged.

liscju commented:
@romani 
1) DONE
```
please open issue on this tool https://github.com/checkstyle/contribution/issues
@attatrol , please take a look .
```
Link to issue https://github.com/checkstyle/contribution/issues/64

2) 
```
It is really strange when maven plugin decide to generate report as single line. @liscju , can you share with me some result of "mvn --version" output ?
```
maven: 3.0.5, java: 1.8.0_45

3) 
```
@liscju , Please do this update. That will let me be ready to merge you even we have problems with reports. 
Please comment out all line in launch.sh from https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch.sh#L95 till the end. Please share reports, do diff by https://www.diffchecker.com/ or similar and share link to diff too. While there is problem in tools we need to do this manually to let this PR be merged.
```

I generated reports before/after for openjdk, guava, spring-framework and hibernate and it looks like there is no regression. I generated diff report with cmd line tool( diff :P) and result of diff is here: http://liscju.github.io/issue2838v2/diffresult.txt , before report: http://liscju.github.io/issue2838v2/before/target/site/checkstyle.html , after report: http://liscju.github.io/issue2838v2/after/target/site/checkstyle.html

4) 

```
Just try to make such example that will take less then 100 symbols in javadoc line.
Now line with value is too long.
Please make example for "An example of how to configure the check for whitespace only around
curly braces is:"
and value="LCURLY, RCURLY"/&gt;
```

Divided line and added example as requested

romani commented:
0) CIs are red - please fix. It is higly recommended to run "mvn clean verify" locally, before pushing code to PR.

1)
> maven: 3.0.5, java: 1.8.0_45

please give all details like , mine are:
```
$ mvn --version
Apache Maven 3.2.5 (12a6b3acb947671f09b81f49094c53f426d8cea1; 2014-12-14T09:29:23-08:00)
Maven home: /opt/apache-maven-3.2.5
Java version: 1.8.0_45, vendor: Oracle Corporation
Java home: /usr/lib/jvm/jdk1.8.0_45/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.2.0-57-generic", arch: "amd64", family: "unix"
```

2)
>
```
 + *               value="ASSIGN,DIV_ASSIGN,PLUS_ASSIGN,MINUS_ASSIGN,STAR_ASSIGN,
+ *                      MOD_ASSIGN,SR_ASSIGN,BSR_ASSIGN,SL_ASSIGN,BXOR_ASSIGN,
+ *                      BOR_ASSIGN,BAND_ASSIGN"/&gt;
```

Hmmm, I was not aware that XML attributes could be multi-lined - http://stackoverflow.com/a/449647/1015848 . Ok.
Configuration has to be ready to use peace of XML code that works out-of-the-box.

romani commented:
merged as FF
